Adding Clause Interleaving to method definitions

Sporarum opened the pull request:
This aims to add the ability to declare functions with many clauses of type parameters, instead of at most one, and to allow those clauses to be interleaved with term clauses:
```scala
def foo[A](x: A)[B](y: B) = (x, y)
```
All user-facing details can be found in the [Scala 3 new features doc](https://github.com/lampepfl/dotty/blob/2a079f92bfc05420e90987d908c41589ac94418d/docs/_docs/reference/other-new-features/generalized-method-syntax.md), and in the [SIP proposal](https://github.com/scala/improvement-proposals/pull/47)

The implementation details are described below in the commit messages

The community's opinion of the feature can be found on [the scala contributors forum](https://contributors.scala-lang.org/t/clause-interweaving-allowing-def-f-t-x-t-u-y-u)
(note however that the description there is somewhat outdated)

### Dependencies

This feature has been [accepted by the SIP committee for implementation](https://github.com/scala/improvement-proposals/pull/47#issuecomment-1287025512), it can therefore become part of the language as an experimental feature at any time.

The feature will be available with `import scala.language.experimental.clauseInterleaving`

### How to make non-experimental
`git revert` the commits named "Make clause interleaving experimental" and "Add import to tests"

### Future Work
1. Implement given aliases with clause interweaving: (to have types depends on terms)
```scala
given myGiven[T](using x: T)[U](using y: U) = (x, y)
```
2. Add interleaved clauses to the left-hand side of extension methods:
```scala
extension (using Int)[A](using A)(a: A)[B](using B)
  def foo: (A, B) = ???
```
3. Investigate usefulness/details of clause interweaving for classes and type currying for types:
```scala
class Foo[A](a: A)[B](b: B)
new Foo(0)("Hello!") // type: Foo[Int][String] ?

type Bar[A][B] = Map[A, B]
Bar[Char] // should this mean [B] =>> Bar[Char][B] ?
```

(Started as a semester project with supervision from @smarter, now part of my missions as an intern at the scala center)

bishabosha commented on docs/docs/reference/other-new-features/generalized-method-syntax.md:
```diff
@@ -0,0 +1,98 @@
+
+### Application
+
+Method application is unchanged, in cases where multiple type clauses are expected but not all are passed, the leftmost ones are inferred.
```
It seems in the following example that the `C` type parameter is the one being inferred, which makes it from the right

```suggestion
Method application is unchanged, in cases where multiple type clauses are expected but not all are passed, the rightmost ones are inferred.
```

Sporarum replied:
Thanks, I added your fix as a commit, when I'll have multiple of those updates/fixes, I'll `rebase -i` to make it clean

soronpo commented:
I'm not sure if you thought about the type system when a method application is split to blocks.
Before this SIP type arguments are inferred as `Nothing` because they all "belong" to the first block.
If this SIP is added, I expect the following to compile:
```scala
  class Box[T](x: T)
  def huge[T1, T2](
    t1: T1, t2 : T2, r : Int
  )[T3, T4](t3: T3, r2: Int, t4: T4)[T5](t5 : T5*): Box[(T1, T2, T3, T4, T5)] = ???

  val h1p1 = huge((1, 2), (3, 4), 5)
  val h1p2 = h1p1((6, 7), 8, (9, 10))
  val h1p3 = h1p2(11, 12)
  val h1pTest: Box[((Int, Int), (Int, Int), (Int, Int), (Int, Int), Int)] = h1p3
```

bishabosha commented:
I think this doc might need updating also https://github.com/lampepfl/dotty/blob/7d322ce0f74cb3b66441fa06a5f47ad9cf89f1b0/docs/_docs/reference/contextual/right-associative-extension-methods.md

Sporarum commented:
> I'm not sure if you thought about the type system when a method application is split to blocks. Before this SIP type arguments are inferred as `Nothing` because they all "belong" to the first block. If this SIP is added, I expect the following to compile:
> 
> ```scala
>   class Box[T](x: T)
>   def huge[T1, T2](
>     t1: T1, t2 : T2, r : Int
>   )[T3, T4](t3: T3, r2: Int, t4: T4)[T5](t5 : T5*): Box[(T1, T2, T3, T4, T5)] = ???
> 
>   val h1p1 = huge((1, 2), (3, 4), 5)
>   val h1p2 = h1p1((6, 7), 8, (9, 10))
>   val h1p3 = h1p2(11, 12)
>   val h1pTest: Box[((Int, Int), (Int, Int), (Int, Int), (Int, Int), Int)] = h1p3
> ```

I share your expectation, sadly this problem is not related specifically to this proposal, but actually more general:
```scala
def small[T](x: T): T = x
val sm1 = small
val res: Int = sm1[Int](1) // method apply in trait Function1 does not take type parameters
```
https://scastie.scala-lang.org/6CfJWHWoSqCDdOc53uWKzA

The above works if you create a function explicitly:
```scala
def small[T](x: T): T = x
val sm1 = [R] => (x: R) => small[R](x)
sm1[Int](1) // 1
```
https://scastie.scala-lang.org/vdIK4Pp3TYOdiPO2oecrEA

Of course we would like the compiler to do this transformation for us, a feature called polymorphic eta-expansion.
I implemented this feature for my semester project: 
https://github.com/lampepfl/dotty/pull/14015
https://contributors.scala-lang.org/t/polymorphic-eta-expansion/5516
The implementation, however, breaks TASTy compatibility, which is why I will be developing a second implementation as part of my internship

You might be able to try both at the same time on [this branch](https://github.com/Sporarum/dotty/tree/interweaving-and-polyEtaExpand), it was however quickly duct-taped together as an experiment, so a lot of edge-cases are not handled correctly

julienrf commented on docs/_docs/reference/other-new-features/generalized-method-syntax.md:
```diff
@@ -0,0 +1,98 @@
+
+### In Scala 3
+
+The new syntax allows any number of type, term and using clause, in any order, optionnaly followed by an implicit clause:
```
The new syntax should be marked as experimental first. It should be available under a compiler option, e.g. `-Yinterleaved-clauses`.

bishabosha replied:
isn't an experimental language import more idiomatic?

soronpo replied:
It shouldn't be needing a special flag, IMO. The feature should just be available if `experimental` is imported.

julienrf commented on tests/neg/namedTypeParams.scala:
```diff
@@ -17,11 +17,13 @@ object Test {
 
   f[X = Int, String](1, "") // error // error
-  f[X = Int][X = Int][Y = String](1, "") // error: illegal repeated type application
+  /* Conflicts with Clause Interweaving, stems from named type parameters assuming one type clause
```
Did you have a look at what it would entail to fix named type arguments?

julienrf replied:
After an offline discussion, I realized that named type arguments are also an experimental feature, and that in most cases you could use interleaved clauses _instead_ of named type arguments, but you are unlikely to use both at the same time.

Sporarum replied:
Since there is a lot of overlap between clause interleaving and named type arguments, and since the latter has apparently been rejected, and should not become part of the language.
Therefore the compatibility between them will not be looked at.

Sporarum commented:
Should interleaved definitions show a special error like "clause interleaving is an experimental feature, you can enable it with ..." ?

julienrf commented:
> Should interleaved definitions show a special error like "clause interleaving is an experimental feature, you can enable it with ..." ?

I don’t think so.

Sporarum commented:
@smarter I will probably rework the commits and update the code slightly, but feel free to review the tests (and more if you feel like it)

Sporarum commented:
I'm in particular having trouble with where to put my test cases, since there are so many existing test files already

bishabosha commented on compiler/src/dotty/tools/dotc/typer/Applications.scala:
```diff
@@ -1148,22 +1148,11 @@ trait Applications extends Compatibility {
-                             |You might also have meant something like ${function}[${args}]""")
+        val function = fun.fun
+        val args = (tree.args ++ fun.args).map(_.show).mkString(", ")
+        errorTree(tree, em"""illegal repeated type application
```
I'm not digging too deep here, but its possible to have repeated type application in general, e.g. curried type lambdas.
```scala
type Foo = [A] =>> [B] =>> Map[A, B]
val x: Foo[String][Int] = Map("abc" -> 123)
```

bishabosha replied:
yes you're right :) (`AppliedTypeTree`)

smarter commented on docs/_docs/reference/other-new-features/generalized-method-syntax.md:
````diff
@@ -0,0 +1,98 @@
+
+In particular, the following does not type check, even though the argument `Char` is only valid for `C`:
+```scala
+def triple[A <: Int][B <: String][C <: Char](a: A, b: B, c: C) = ???
````
This example should be changed now that we don't allow repeated type applications, right?

smarter commented on project/MiMaFilters.scala:
```diff
@@ -3,6 +3,12 @@ import com.typesafe.tools.mima.core._
 object MiMaFilters {
   val Library: Seq[ProblemFilter] = Seq(
+    
+    //  New experimental features in 3.2.X
```
We're now in 3.3.x

Sporarum commented:
I reworked the commits, I think this PR is in a really good shape !

The commit should be modular enough that going commit by commit should make understanding it easier

So I invite you to review it
(In particular @smarter)

PS: The "Make clause interleaving experimental" should be self-contained enough to illustrate well how to make something experimental

smarter commented on compiler/src/dotty/tools/dotc/parsing/Parsers.scala:
```diff
@@ -3079,6 +3079,67 @@ object Parsers {
+     *                   | DefTermParamClause 
+     *                   | UsingParamClause
+     */
+    def typeOrTermParamClause(
```
This method appears to be unused

Sporarum replied:
It indeed is, I left it for completeness, should I remove it ?

smarter commented on compiler/src/dotty/tools/dotc/parsing/Parsers.scala:
```diff
@@ -3079,6 +3079,67 @@ object Parsers {
+     */
+    def typeOrTermParamClauses(
+                    ownerKind: ParamOwner,
+                    ofClass: Boolean = false,
+                    ofCaseClass: Boolean = false,
+                    givenOnly: Boolean = false,
```
There is only one call to this method and it doesn't specify these arguments explicitly.

Sporarum replied:
I left them to have the same signature as the other, should I simplify it instead ?

smarter commented on compiler/src/dotty/tools/dotc/typer/Applications.scala:
```diff
@@ -443,17 +443,10 @@ trait Applications extends Compatibility {
-            rec(instantiateWithTypeVars(funType))
-          case tp => tp
-        }
-      }
```
Even if this isn't currently needed with the restriction on type parameter clauses, I don't think it's a bad idea to keep the more general version of this code.

Sporarum commented:
The failing Scaladoc test will be fixed once I rebase on main
(files contain "... 2022" instead of "... 2023", happy new year !)

Sporarum commented:
As all dependencies have been merged with main, this PR can be merged as soon as you think it is good enough !

@smarter did I correctly address your comments ? Is there anything else that needs changing ?

Also should someone else give it a look, @sjrd maybe ?

soronpo commented:
IMO at least some of the neg tests should include a `.check` to regression tests against the error messages themselves.

soronpo commented:
Also I think we need extension method interactions in the tests as well.

sjrd commented on compiler/src/dotty/tools/dotc/typer/Applications.scala:
```diff
@@ -1143,8 +1150,12 @@ trait Applications extends Compatibility {
-        errorTree(tree, em"illegal repeated type application")
+      case fun: TypeApply if !ctx.isAfterTyper =>
+        val function = fun.fun
+        val args = (tree.args ++ fun.args).map(_.show).mkString(", ")
```
Shouldn't it be the other way around?
```suggestion
        val args = (fun.args ++ tree.args).map(_.show).mkString(", ")
```
Is there a test for this?

sjrd commented on docs/_docs/reference/other-new-features/generalized-method-syntax.md:
````diff
@@ -0,0 +1,95 @@
+The old syntax only allowed zero or one type parameter clause, followed by any number of term clauses, optionnally followed by an implicit clause:
+
+```scala
+def foo[T, U](x: T)(y: U)(z: Int, s: String)(using Ord[Int])(a: Array[T])(implicit l: List[U])
````
The section title is "In Scala 2" but there's a `using` clause here. That's not consistent.

sjrd commented on docs/_docs/reference/other-new-features/generalized-method-syntax.md:
```diff
@@ -0,0 +1,95 @@
+movedTo: https://docs.scala-lang.org/scala3/reference/other-new-features/generalized-method-syntax.html
+---
+
+The inclusion of using clauses is not the only way in which methods have been updated, type parameter clauses are now allowed in any number and at any position.
```
I guess there should be some "experimental" marker at the top of this page.

sjrd commented on docs/_docs/reference/other-new-features/generalized-method-syntax.md:
```diff
@@ -0,0 +1,95 @@
+
+### Dependent Type Clauses
+
+As type clauses can come after term clauses, it is now possible to have type parameters depend on term parameters:
```
```suggestion
As type clauses can come after term clauses, it is now possible to have type parameters that depend on term parameters:
```

sjrd commented on docs/_docs/reference/other-new-features/generalized-method-syntax.md:
````diff
@@ -0,0 +1,95 @@
+```scala
+trait Key { type Value }
+trait DB {
+  def get(k: Key): Option[k.Value] // dependent return type
````
Being pedantic:
```suggestion
  def get(k: Key): Option[k.Value] // dependent result type
```
Java has return types. Scala has result types.

sjrd commented on docs/_docs/reference/other-new-features/generalized-method-syntax.md:
````diff
@@ -0,0 +1,95 @@
+```
+
+Note that simply replacing `V` by `k.Value` would not be equivalent. For example, if `k.Value` is `Int`, the above allows: 
+`getOrElse(k)[Number](0.1)`, which returns a `Number`.
````
`Number` is not a supertype of `Int`. It's only a supertype of `Integer`. You need a different example here.

Sporarum replied:
I think I'll use `Option` then, thank you for reminding me

sjrd commented on docs/_docs/reference/other-new-features/generalized-method-syntax.md:
```diff
@@ -0,0 +1,95 @@
+
+### Application
+
+Method application is unchanged, in cases where multiple type clauses are expected but not all are passed, the rightmost ones are inferred.
```
```suggestion
Method application is unchanged.
When multiple type clauses are expected but not all are passed, the rightmost ones are inferred.
```

sjrd commented on docs/_docs/reference/other-new-features/generalized-method-syntax.md:
```diff
@@ -0,0 +1,95 @@
+
+### Extension Methods
+
+Naturally extension methods follow the same syntax, so the following is valid:
```
```suggestion
Extension methods follow the same syntax, so the following is valid:
```

sjrd commented on docs/_docs/reference/other-new-features/generalized-method-syntax.md:
```diff
@@ -0,0 +1,95 @@
+
+We recommand to always put a unique type clause at the beginning, unless it is not possible to do so.
+For example, the extension method `zipWith` above should be written `zipWith[U, V](l2: List[U], l3: List[V]): List[(T,U,V)]` instead.
+The `getOrElse` method on the other hand can not be written with a leading type clause, it is therefore as recommended as is.
```
```suggestion
The `getOrElse` method on the other hand can not be written with a leading type clause.
```

sjrd commented on project/MiMaFilters.scala:
```diff
@@ -14,12 +15,16 @@ object MiMaFilters {
     ProblemFilters.exclude[MissingFieldProblem]("scala.runtime.stdLibPatches.language#experimental.captureChecking"),
     ProblemFilters.exclude[MissingClassProblem]("scala.runtime.stdLibPatches.language$experimental$pureFunctions$"),
     ProblemFilters.exclude[MissingClassProblem]("scala.runtime.stdLibPatches.language$experimental$captureChecking$"),
+    // /New experimental features in 3.3.X
```
```suggestion
    // New experimental features in 3.3.X
```
But perhaps it's already redundant with the title above?

Sporarum replied:
```suggestion
    // End of new experimental features in 3.3.X
```

Is this better ?
The idea is to clearly mark both the beginning and end of the experimental features
The entries below this comment are no longer experimental features

sjrd replied:
Oh. Er, yes, definitely. That is not at all how I had interpreted the `/`.

sjrd commented on tests/neg/interleaving-ab.scala:
```diff
@@ -0,0 +1,11 @@
+import scala.language.experimental.clauseInterleaving
+
+object Ab:
+    given String = ""
```
```suggestion
object Ab:
  given String = ""
```
Same in many of the test files.

sjrd replied:
Also: missing EOL at EOF, as reported by git and GitHub. Also same in many of the test files.

sjrd commented on tests/neg/interleaving-nameCollision.scala:
```diff
@@ -0,0 +1,5 @@
+
+object nameCollision:
+    def f[T](x: T)[U](y: U) = (x,y)
+    def f[T](x: T, y: T) = (x,y) // error
```
Is this a name collision, or a signature collision (after erasure)?

sjrd commented on tests/neg/namedTypeParams.scala:
```diff
@@ -24,4 +24,10 @@ object Test {
   f[Y = String][X = Int](1, "") // error: illegal repeated type application
   f[Y = String][Int](1, "") // error: illegal repeated type application
+
+  import scala.language.experimental.clauseInterleaving
```
If you want this `import` to apply only locally, it's best to explicitly create a separate scope.

sjrd commented on tests/pos/interleaving-functor.scala:
```diff
@@ -0,0 +1,19 @@
+object functorInterweaving:
+  //taken from https://dotty.epfl.ch/docs/reference/contextual/type-classes.html
+  //at version 3.1.1-RC1-bin-20210930-01f040b-NIGHTLY
+  //modified to have type interveawing
```
```suggestion
  //modified to have type interleaving
```

sjrd commented on tests/pos/interleaving-nameCollision.scala:
```diff
@@ -0,0 +1,6 @@
+import scala.language.experimental.clauseInterleaving
+import scala.annotation.targetName
+
+object nameCollision:
```
Again, this is probably more about signature collision than name collision.

sjrd reviewed (changes requested):
Mostly nitpicks and style. Some (I think) mistakes in the docs.

Regarding tests: we really need some `run` tests in addition to the `pos` tests. Currently there's nothing checking that the bytecode produced for these things is viable.

We should also have some tests for Java interop: how does Java see methods defined in this way? What's their generic signature?

Sporarum commented on compiler/src/dotty/tools/dotc/parsing/Parsers.scala:
```diff
@@ -3079,6 +3079,42 @@ object Parsers {
+     *                          | DefTermParamClause 
+     *                          | UsingParamClause
+     */
+    def typeOrTermParamClauses(
+                    ownerKind: ParamOwner,
+                    numLeadParams: Int = 0
+                   ): List[List[TypeDef] | List[ValDef]] =
```
I copied the indentation of `termParamClause`, I have no preference either way, but it's true that uniformity would be better,
What should I do then ?

sjrd replied:
When it was named `paramClause`, it looked like
```scala
    def paramClause(
                    ownerKind: ParamOwner,
                    numLeadParams: Int = 0
                   ): List[List[TypeDef] | List[ValDef]] =
```
which is a defensible, although rare, indentation style: the `)` aligns with the `(`, and the params align with that + 1 space.

Now that the method has changed *name*, the `(` moved, so nothing is aligned anymore. You could add 11 more spaces to the 3 param lines, to fix it, but I suggest instead to take the opportunity to switch to a more widespread style, as follows:
```scala
    def typeOrTermParamClauses(
      ownerKind: ParamOwner,
      numLeadParams: Int = 0
    ): List[List[TypeDef] | List[ValDef]] =
```
That style does not depend on the *length* of the method name.

Sporarum replied:
Okay thank you

Sporarum replied:
Wouldn't it be:
```scala
    def typeOrTermParamClauses(
        ownerKind: ParamOwner,
        numLeadParams: Int = 0
    ): List[List[TypeDef] | List[ValDef]] =
```
or
```scala
    def typeOrTermParamClauses(
        ownerKind: ParamOwner,
        numLeadParams: Int = 0): List[List[TypeDef] | List[ValDef]] =
```
Given `infixOps` is:
```scala
    def infixOps(
        first: Tree, canStartOperand: Token => Boolean, operand: Location => Tree,
        location: Location,
        kind: ParseKind,
        isOperator: => Boolean): Tree =
```

sjrd replied:
Ah yes, indeed.

Sporarum commented on compiler/src/dotty/tools/dotc/parsing/Parsers.scala:
```diff
@@ -3139,21 +3175,25 @@ object Parsers {
      *  @return   the list of parameter definitions
      */
-    def paramClause(nparams: Int,                            // number of parameters preceding this clause
+    def termParamClause(nparams: Int,                            // number of parameters preceding this clause
                     ofClass: Boolean = false,                // owner is a class
```
I'm not sure I understand what you mean, how would the indentation change ?

sjrd replied:
Same story as my other comment.

Sporarum commented:
@sjrd I believe I have addressed your comments, I didn't rebase yet so you can more easily see what I did to address them.
I have only put one run test, it is relatively thorough, but I wouldn't mind some guidance if you think more is needed

@smarter Could you give it another look, I think I changed what you asked me too, but I need your confirmation

Sporarum commented:
Note to myself:
I need to rollback the reference/syntax.md in the "make experimental" commit

sjrd commented:
Looks fine for me except for the Java interop test we discussed off-line.

Sporarum commented:
Testing the java interop lead me to discover it does not work for extension methods either, see https://github.com/lampepfl/dotty/issues/16727

As such I believe java interop to be outside of the scope of this PR

Sporarum commented on scaladoc/src/dotty/tools/scaladoc/tasty/ClassLikeSupport.scala:
```diff
@@ -12,6 +12,9 @@ import NameNormalizer._
 import SyntheticsSupport._
 import dotty.tools.dotc.core.NameKinds
 
+// Please use this only for things defined in the api.scala file
```
There were two problems with the scladoc:
* An incorrect heuristic for spliting the rhs from the lhs of extension methods
* Was not designed to accomodate more than 2 type clauses

The other PR fixes the first point, as it impacts already valid code
The second point is addressed in this PR, as it is unnecessary for code without interleaving

smarter commented on tests/neg/extension-methods.scala:
```diff
@@ -15,4 +15,23 @@ object Test {
+
+  extension [C1 <: Chain](c1: C1)
+    def ?:[C2 <: c1.Tail](c2: C2): c2.Tail = ??? // Not found: c1
+
```
I'm also wondering if we ever discussed allowing type parameter clauses after the extension term parameter clause:
```scala
extension (c1: Chain)[C2 <: c1.Tail]
  def ...
```

Sporarum replied:
I believe we deemed it outside the scope of the proposal, in particular here is what's written in the [sip](https://github.com/scala/improvement-proposals/blob/main/content/clause-interleaving.md#lhs-of-extension-methods)

> The left hand side of extension methods remains unchanged, since they only have one explicit term clause, and since the type parameters are very rarely passed explicitly, it is not as necessary to have multiple type clauses there.

It also complicates even more speaking about the shuffling logic for right-associative extension methods (an out-of-date explanation is [here](https://docs.scala-lang.org/scala3/reference/contextual/right-associative-extension-methods.html))

smarter reviewed (commented):
Otherwise LGTM.

Sporarum commented on tests/neg/extension-methods.scala:
```diff
@@ -15,4 +15,23 @@ object Test {
+  extension [C2 <: c1.Tail](c2: C2) // Not found: c1
+    def !:[C1 <: Chain](c1: C1): c2.Tail = ???
+
+  extension [C1 <: Chain](c1: C1)
+    def ?:[C2 <: c1.Tail](c2: C2): c2.Tail = ??? // Not found: c1
```
We didn't discuss it, but it seems to be the intended behavior (even before this feature):
https://scastie.scala-lang.org/A6pbaHSuQNucDsrOALxQjw

The reason I believe is that right associative extension methods get their parameters shuffled:
* We can probably allow the first line simply, but it looks really confusing, because things depend on parameter that come after in the definition
* The second looks normal, but is actually translated to `[C2 <: c1.Tail](c2: C2)[C1 <: Chain](c1: C1): c2.Tail`, which again has weird references

No matter what we do, the result is confusing, so not allowing either might be the best option

The thing that is missing is a friendlier error message, as it is not obvious why the second it not allowed to a user not familiar with the internals
There are also shadowing surprises like:
```Scala
val c1: Chain = "Looks shadowed, but isn't".asInstanceOf[Chain]
extension [C1 <: Chain](c1: C1)
  def ?:[C2 <: c1.Tail](c2: C2): c2.Tail = ??? // c1 is the val above, not the parameter !
```

smarter reviewed (approved):
:sparkle:
